docs/user/reference/action.md

```diff
@@ -18,4 +18,49 @@ https://discourse.charmhub.io/t/new-feature-in-juju-2-8-improved-actions-experie
 
 (Starting with `juju v.3.0`: ) The execution of an action is organised into {ref}`tasks <task>` and {ref}`operations <operation>`. (If an action defines a named unit of work -- e.g., back up the database -- that can be executed on selected units, a task is the execution of the action on each target unit, and an operation is the group of tasks queued by running an action across one or more units.)
 
+(action-execution)=
+## Action execution
 
+Actions operate in an execution environment similar to a {ref}`hook-execution hook`, with additional environment variables available:
+
+* JUJU_ACTION_NAME holds the name of the action.
+* JUJU_ACTION_UUID holds the UUID of the action.
+
+An action is used to perform a named, parameterised operation and report back the results of said operation.
+Actions are defined by the charm and invoked by the user (using {ref}`command-juju-run`). The default behaviour is that
+the command blocks and wait for the action to complete. During this time any progress messages as reported by the action
+are logged to the terminal. When the action completes, the result is printed.
+An action result is a map of key values, containing data set by the action as it runs, plus the overall exit code
+of the action process itself, and the content of stdout and stderr.
+
+The code used to implement an action can call the following hook commands in addition to {ref}`list-of-hook-commands the others`:
+* action-log: to report a progress message
+* action-get: to get the value of a named action parameter as supplied by the user
+* action-set: to set a value in the action results map
+* action-fail: to mark the action as failed along with a failure message
+
+In most cases, an action only has a need to run hook commands such as `config-get` to supplement the configuration passed
+in via the action parameters. A action may also commonly use `status-set` to update the unit or application status while
+it is running.
+
+### What triggers it?
+
+A charm user invoking the action name from the Juju CLI (`juju run <unit/0> foo`,  `juju run <unit/0> <unit/1> foo`).
+
+### Who gets it?
+
+All the units that the charm user has run the action on.
+
+```{note}
+When implementing a hook using {ref}`Ops <ops-ops>`, any hyphens in action names are replaced with underscores
+in the corresponding event names.
+For example, an action named `snapshot-database` would result in an event named `snapshot_database_action`
+being triggered when the action is invoked.
+```
+
+```{tip}
+The action handler in the charm can in principle cause other events to be fired.
+For example:
+ - deferred events will trigger before the action.
+ - if the action handler updates relation data, a `<endpoint name>_relation_changed` will be emitted afterwards on the affected units.
```

Comment:
```suggestion
 - if the action handler updates relation data, a `<endpoint name>_relation_changed` hook will be emitted afterwards on the affected units.
```
I don't have the anchor at hand, but we should probably link to the relevant  hook section too.